csrc/preseg_passes/remove_bcast_squeeze.cpp

```diff
@@ -165,51 +154,167 @@ std::vector<bool> nonPreservedDims(const AxisOps& ops) {
+  // _intermediate_ tensor between prev and next. This means there is one main
+  // loop iteration per "aligned" pair of ops.
+  //
+  // In each main loop iteration, we first process _unaligned_ ops in next and
+  // prev, i.e. previous squeezes and next broadcasts. We cancel as many of
+  // these operations as possible by converting them to PRESERVE ops and we
+  // insert the remainder unmodified. After processing unaligned ops, we know
+  // the next op pair is an aligned pair, meaning the previous op in the pair
+  // results in an intermediate axis that is consumed in the next op. This
+  // implies that the previous op is not a SQUEEZE and the next op is not a
+  // BROADCAST. If either is PRESERVE, then we can simplify the composition by
+  // simply using the other op. The only case left is then "broadcast then
+  // squeeze", which we ignore since it does not consume an input axis or
+  // produce an output axis.
+  //
+  // For example:
+  //
+  //   prev = [ S S P B P P S ]
+  //   next = [ B P P S P B B ]
+  //
+  // The input has dimension 6, the intermediate tensor has dimension 4, and the
+  // output has dimension 6. The composition is equivalent to [ S P P P B ]. We
+  // perform 4 main loop iterations corresponding to the four intermediate
+  // dimensions.
+  //
+  //   ops = []
+  //   prev = [ (prev_pos)S S P B P P S ]
+  //   next = [ (next_pos)B P P S P B B ]
+  //
+  //   First main loop iteration:
+  //
+  //     (run processUnalignedOps()):
+  //
+  //     // one cancelled SQUEEZE becomes PRESERVE
+  //     ops = [ S P ]
+  //     prev = [ S S (prev_pos)P B P P S ]
+  //     next = [ B (next_pos)P P S P B B ]
+  //
+  //     (process aligned ops "preserve then preserve")
+  //     ops = [ S P P ]
+  //     prev = [ S S P (prev_pos)B P P S ]
+  //     next = [ B P (next_pos)P S P B B ]
+  //
+  //   Second main loop iteration:
+  //
+  //     (run processUnalignedOps()) // no change
+  //
+  //     (process aligned ops "broadcast then preserve")
+  //     ops = [ S P P B ]
+  //     prev = [ S S P B (prev_pos)P P S ]
+  //     next = [ B P P (next_pos)S P B B ]
+  //
+  //   Third main loop iteration:
+  //     (run processUnalignedOps()) // no change
+  //
+  //     (process aligned ops "preserve then squeeze")
+  //     // Note that we would normally insert SQUEEZE resulting in
+  //        ops = [ S P P B S ], but instead we remove the
+  //        previously-inserted BROADCAST to PRESERVE in pushOp(AxisOp::SQUEEZE)
+  //     ops = [ S P P ]
+  //     prev = [ S S P B P (prev_pos)P S ]
+  //     next = [ B P P S (next_pos)P B B ]
+  //
+  //   Third main loop iteration:
+  //     (run processUnalignedOps()) // no change
+  //
+  //     (process aligned ops "preserve then preserve")
+  //     ops = [ S P P P ]
+  //     prev = [ S S P B P P (prev_pos)S ]
+  //     next = [ B P P S P (next_pos)B B ]
+  //
+  //   Post-processing:
+  //     (run processUnalignedOps())
+  //     ops = [ S P P P B ]
```

Comment:
seems should be `ops = [ S P P P P B ]`, where the last `P` comes from `S` in `prev` and `B` in `next`

Reply:
Oh you're right that this is wrong.  The output also should be 6D... Let me take another shot at this. I'll ping you when it's ready for another look.

Reply:
I found the error. When I push a squeeze onto `ops` and `ops.back() == SQUEEZE`, we should convert `ops.back()` to `PRESERVE` instead of removing it. This is now fixed.